### PR TITLE
SALTO-2591: add deploy capabilities for sections in zendesk guid

### PR DESCRIPTION
**Pull request by edenhassid**
_add deploy capabilities for sections in zendesk guid_

---

_Additional context for reviewer_
None

---
_Release Notes_: 
None

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/zendesk-adapter/src/adapter.ts [hunk 1]
```diff
@@ -79,6 +79,8 @@ import removeBrandLogoFieldFilter from './filters/remove_brand_logo_field'
 import { getConfigFromConfigChanges } from './config_change'
 import { dependencyChanger } from './dependency_changers'
 import customFieldOptionsFilter from './filters/add_restriction'
+import hcSectionFilter from './filters/help_center_section'
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:82]: terminology nit: guide and help center are the same, to avoid confusion I think we should stick to one term in the code (and so far we've used guide afaict). I think we can just have a `guide` folder under `filters` and put all the guide-related filters there

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+} from '@salto-io/adapter-api'
+import _ from 'lodash'
+import Joi from 'joi'
+import { createSchemeGuard } from '@salto-io/adapter-utils'
+import { FilterCreator } from '../filter'
+import { deployChange, deployChanges } from '../deployment'
+
+const SECTION_TYPE_NAME = 'section'
+
+type TranslationType = InstanceElement &{
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:31]: tiny nit:
```suggestion
type TranslationType = InstanceElement & {
```

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    title: string
+    body?: string
+    locale: string
+}
+
+export type sectionType = InstanceElement & {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:37]: nit:
```suggestion
export type SectionType = InstanceElement & {
```

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  locale: string
+}
+
+const TRANSLATION_SCHEMA = Joi.object({
+  locale: Joi.string().required(),
+  body: Joi.string(),
+  title: Joi.string().required(),
+}).unknown(true).required()
+
+export const isTranslation = createSchemeGuard<TranslationType>( // TODO
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:47]: I'm assuming this will be handled before merging :)

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  TRANSLATION_SCHEMA, 'Received an invalid value for translation'
+)
+
+const addTranslationValues = (change: Change<InstanceElement>): void => {
+  const currentLocale = getChangeData(change).value.source_locale
+  const translation = getChangeData(change).value.translations
+    .filter(isTranslation)
+    .find((tran:TranslationType) => tran.locale === currentLocale)
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:55]:
```suggestion
    .find((tran: TranslationType) => tran.locale === currentLocale)
```
also - I think we will want to convert the locales to references eventually and then this will need to change (but we can change that later as well)

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  if (translation) {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:56]: nit: the convention is to explicitly compare to `undefined`, to avoid mistakes (e.g. empty strings) 
```suggestion
  if (translation !== undefined) {
```

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    getChangeData(change).value.name = translation.title
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:57]: I see we're doing this in a few other places, but @ori-moisis just making sure it's ok to change on the data itself without cloning?

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  locale: Joi.string().required(),
+  body: Joi.string(),
+  title: Joi.string().required(),
+}).unknown(true).required()
+
+export const isTranslation = createSchemeGuard<TranslationType>( // TODO
+  TRANSLATION_SCHEMA, 'Received an invalid value for translation'
+)
+
+const addTranslationValues = (change: Change<InstanceElement>): void => {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:51]: please add a docstring explaining why this is needed (the data duplication that we handle on fetch)

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  const translation = getChangeData(change).value.translations
+    .filter(isTranslation)
+    .find((tran:TranslationType) => tran.locale === currentLocale)
+  if (translation) {
+    getChangeData(change).value.name = translation.title
+    getChangeData(change).value.description = translation.body ?? ''
+  }
+}
+
+const filterCreator: FilterCreator = ({ client, config }) => ({
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:62]: please add a docstring

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    )
+    return { deployResult, leftoverChanges }
+  },
+  onDeploy: async (changes: Change<InstanceElement>[]): Promise<void> => {
+    changes
+      .filter(change => getChangeData(change).elemID.typeName === SECTION_TYPE_NAME)
+      .forEach(change => {
+        delete getChangeData(change).value.name
+        delete getChangeData(change).value.description
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:95]: to make sure we maintain consistency between `onFetch` and `onDeploy` in the future, maybe move this to a helper function that receives a `Section` instance and removes the fields from it? then we can use this function in both flows

File: packages/zendesk-adapter/src/filters/help_center_section_translation.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  const currentLocale = data.value.locale
+  const parents = getParents(data)
+  return parents.find(parent => parent.source_locale?.value.value.id === currentLocale) ?? false
+}
+
+/**
+ * a different filter is needed for the default translation as it is deployed during the deployment
+ * of the section itself. Therefor, this filter removes the default translation from the
+ * leftover changes.
+ */
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/help_center_section_translation.ts:39]: 👌 
nit: maybe make the explanation slightly more about the behavior and less about the implementation, e.g.
```suggestion
 * a different filter is needed for the default translation as it is deployed during the deployment
 * of the section itself. Therefore, this filter marks the default translation as successfully deployed without actually deploying it. The rest of the translations will be deployed in <...>
 */
```

File: packages/zendesk-adapter/test/filters/help_center_section.test.ts [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        source_locale: 'he',
+        translations: [
+          sectionTranslationInstance.value,
+        ],
+      })
+    })
+  })
+
+  describe('preDeploy', () => {
+    it('should add the name and description fields after deploy', async () => {
```
Comment (netama) [inline packages/zendesk-adapter/test/filters/help_center_section.test.ts:95]: nit: should be `before`?

File: packages/zendesk-adapter/test/filters/help_center_section_translation.test.ts [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        helpCenterLocaleType.elemID.createNestedID('instance', 'Test1'), helpCenterLocaleInstance
+      ),
+      translations: [
+        heSectionTranslationInstance.value,
+        enSectionTranslationInstance.value,
+      ],
+    }
+  )
+
+  heSectionTranslationInstance.annotations[CORE_ANNOTATIONS.PARENT] = [sectionInstance.value]
```
Comment (netama) [inline packages/zendesk-adapter/test/filters/help_center_section_translation.test.ts:92]: the parent should be a reference expression, not a value, no? what am I missing?

File: packages/zendesk-adapter/test/filters/help_center_section_translation.test.ts [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+    'instance',
+    sectionType,
+    {
+      name: 'name',
+      description: 'description',
+      source_locale: new ReferenceExpression(
+        helpCenterLocaleType.elemID.createNestedID('instance', 'Test1'), helpCenterLocaleInstance
+      ),
+      translations: [
+        heSectionTranslationInstance.value,
```
Comment (netama) [inline packages/zendesk-adapter/test/filters/help_center_section_translation.test.ts:86]: shouldn't these be reference expressions?

**Review by netama — COMMENTED** [review]
great progress :) I think we have a few more pieces to add before we can merge this.
specifically:
1. we should have change validators to ensure that we actually have a translation for the default locale - we can add these in a different PR if you prefer, but I think we should merge them at the same time
2. we should make sure if the default translation is the only thing that changed, we do deploy that update and don't omit it
3. we should add more tests (including an e2e one if possible) for edge cases

File: packages/zendesk-adapter/src/filters/help_center_section.ts [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        delete obj.value.name
+        delete obj.value.description
+      })
+  },
+  preDeploy: async (changes: Change<InstanceElement>[]): Promise<void> => {
+    changes
+      .filter(change => getChangeData(change).elemID.typeName === SECTION_TYPE_NAME)
+      .forEach(addTranslationValues)
+  },
+  deploy: async (changes: Change<InstanceElement>[]) => {
```
Comment (noamhammer) [inline packages/zendesk-adapter/src/filters/help_center_section.ts:77]: Not sure where to put the comment but I will put it here - You don't handle the case for removal, if you would remove a section with all its translations, you will fail to remove the translations. You need to ignore the removal changes on the translations if the section was removal

**Comment by noamhammer** [comment]
@edenhassid , I believe that you can easily expand those filters to support other types. It might be valuable for categories and also articles

File: packages/zendesk-adapter/test/filters/help_center_section_translation.test.ts (hunk 13 above)
Comment (edenhassid) [reply packages/zendesk-adapter/test/filters/help_center_section_translation.test.ts:92]: I checked and here for example the parent is not a reference expression.
image

File: packages/zendesk-adapter/src/adapter.ts [hunk 16]
```diff
@@ -458,7 +466,7 @@ export default class ZendeskAdapter implements AdapterOperations {
           subdomainToGuideChanges[subdomain]
         )
         const guideChangesBeforeRestore = [...brandDeployResults.appliedChanges]
-        await runner.onDeploy(appliedChangesBeforeRestore)
+        await runner.onDeploy(guideChangesBeforeRestore)
```
Comment (netama) [inline packages/zendesk-adapter/src/adapter.ts:469]: good catch!
@Eldardar fyi as well in case you run into something similar in your work

File: packages/zendesk-adapter/src/change_validators/translation_for_default_locale.ts [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    return false
+  }
+  const sourceLocale = instance.value.source_locale
+  const translation = instance.value.translations
+    .filter(isTranslation)
+    .find(tran => tran.locale.value.value.id === sourceLocale)
+  return (translation === undefined) // no translation for the source_locale
+}
+
+export const translationForDefaultLocaleValidator: ChangeValidator = async changes => {
```
Comment (netama) [inline packages/zendesk-adapter/src/change_validators/translation_for_default_locale.ts:61]: nit: let's add documentation (I see you documented the inner function above - if this was intended as the validator's documentation please just move it :) )

File: packages/zendesk-adapter/src/change_validators/translation_for_default_locale.ts [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    .map(data => resolveValues(data, lookupFunc))
+    .filter(noTranslationForDefaultLocale)
+    .toArray()
+
+  return relevantInstances
+    .flatMap(instance => [{
+      elemID: instance.elemID,
+      severity: 'Error',
+      message: `Instance ${instance.elemID.getFullName()} does not have a translation for the source locale`,
+      detailedMessage: `Instance ${instance.elemID.getFullName()} does not have a
```
Comment (netama) [inline packages/zendesk-adapter/src/change_validators/translation_for_default_locale.ts:76]: nit: it's better if this is actionable, maybe something like
```suggestion
      detailedMessage: `${instance.elemID.type} instance "${instance.elemID.name}" must have a 
```

File: packages/zendesk-adapter/src/change_validators/translation_for_default_locale.ts [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    .filter(instance => PARENTS_TYPE_NAMES.includes(instance.elemID.typeName))
+    .map(data => resolveValues(data, lookupFunc))
+    .filter(noTranslationForDefaultLocale)
+    .toArray()
+
+  return relevantInstances
+    .flatMap(instance => [{
+      elemID: instance.elemID,
+      severity: 'Error',
+      message: `Instance ${instance.elemID.getFullName()} does not have a translation for the source locale`,
```
Comment (netama) [inline packages/zendesk-adapter/src/change_validators/translation_for_default_locale.ts:75]: we're going to group all relevant instances by the message, so it's better not to have the instance id in it. it is useful to have the type though, so maybe:
```suggestion
      message: `${instance.elemID.type} instance does not have a translation for the source locale`,
```

File: packages/zendesk-adapter/src/filters/fetch_section_and_category.ts [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+import { FilterCreator } from '../filter'
+
+const PARENTS_TYPE_NAMES = ['section', 'category']
+
+const removeNameAndDescription = (elem: InstanceElement): void => {
+  delete elem.value.name
+  delete elem.value.description
+}
+
+const filterCreator: FilterCreator = () => ({
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/fetch_section_and_category.ts:26]: nits: 
1. please add documentation here as well - I see you already explained the need in help_center_section_and_category.ts, so please just link to it and explain why this is done separtely.
2. maybe give the filters similar names? (one starts with `help_center_` and the other doesn't)

File: packages/zendesk-adapter/src/filters/fetch_section_and_category.ts [hunk 21]
```diff
@@ -0,0 +1,34 @@
...
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import { Element, InstanceElement, isInstanceElement } from '@salto-io/adapter-api'
+import { FilterCreator } from '../filter'
+
+const PARENTS_TYPE_NAMES = ['section', 'category']
+
+const removeNameAndDescription = (elem: InstanceElement): void => {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/fetch_section_and_category.ts:21]: please use a shared function with the other filter running this `onDeploy`

same for `PARENTS_TYPE_NAMES` (and if it will be exported, maybe call it something more specific like `TRANSLATION_PARENT_TYPE_NAMES`?)

**Review by netama — APPROVED** [review]
⭐
